src/controller/python/ChipDeviceController-ScriptBinding.cpp

```diff
@@ -181,14 +183,78 @@ uint64_t pychip_GetCommandSenderHandle(chip::DeviceProxy * device);
 
 chip::ChipError::StorageType pychip_InteractionModel_ShutdownSubscription(uint64_t subscriptionId);
+
+//
+// Storage
+//
+void pychip_Storage_InitializeStorageAdapter(chip::Controller::Python::PyObject * context,
+                                             chip::Controller::Python::SyncSetKeyValueCb setCb,
+                                             chip::Controller::Python::SetGetKeyValueCb getCb,
+                                             chip::Controller::Python::SyncDeleteKeyValueCb deleteCb);
+void pychip_Storage_ShutdownAdapter();
+}
+
+void pychip_Storage_InitializeStorageAdapter(chip::Controller::Python::PyObject * context,
+                                             chip::Controller::Python::SyncSetKeyValueCb setCb,
+                                             chip::Controller::Python::SetGetKeyValueCb getCb,
+                                             chip::Controller::Python::SyncDeleteKeyValueCb deleteCb)
+{
+    sStorageAdapter = new chip::Controller::Python::StorageAdapter(context, setCb, getCb, deleteCb);
+}
+
+void pychip_Storage_ShutdownAdapter()
+{
+    delete sStorageAdapter;
+}
+
+chip::Controller::Python::StorageAdapter * pychip_Storage_GetStorageAdapter()
+{
+    return sStorageAdapter;
+}
+
+ChipError::StorageType pychip_DeviceController_StackInit()
+{
+    CHIP_ERROR err;
+
+    VerifyOrDie(sStorageAdapter != nullptr);
+
+    err = sFabricStorage.Initialize(sStorageAdapter);
+    VerifyOrReturnError(err == CHIP_NO_ERROR, err.AsInteger());
+
+    FactoryInitParams factoryParams;
+    factoryParams.fabricStorage = &sFabricStorage;
+    factoryParams.imDelegate    = nullptr;
+
+    ReturnErrorOnFailure(DeviceControllerFactory::GetInstance().Init(factoryParams).AsInteger());
+
+    //
+    // In situations where all the controller instances get shutdown, the entire stack is then also
+    // implicitly shutdown. In the REPL, users can create such a situation by manually shutting down
+    // controllers (for example, when they call ChipReplStartup::LoadFabricAdmins multiple times). In
+    // that situation, momentarily, the stack gets de-initialized. This results in further interactions with
+    // the stack being dangerous (and in fact, causes crashes).
+    //
+    // This retain call ensures the stack doesn't get de-initialized in the REPL.
+    //
+    DeviceControllerFactory::GetInstance().RetainSystemState();
+
+#if CHIP_DEVICE_CONFIG_ENABLE_COMMISSIONER_DISCOVERY
+    chip::app::DnssdServer::Instance().StartServer(chip::Dnssd::CommissioningMode::kDisabled);
+#endif // CHIP_DEVICE_CONFIG_ENABLE_COMMISSIONER_DISCOVERY
+
+    return CHIP_NO_ERROR.AsInteger();
 }
 
 ChipError::StorageType pychip_DeviceController_NewDeviceController(chip::Controller::DeviceCommissioner ** outDevCtrl,
                                                                    chip::NodeId localDeviceId)
 {
+    static uint16_t sFabricIndex = 1;
+
     *outDevCtrl = new chip::Controller::DeviceCommissioner();
```

Comment:
Realized any error below could lead to memory leak, I guess this could be resolved with the comment below in another PR?